### ISSUE TITLE
Turbo-frame request variant

### ISSUE THREAD

**Issue by tleish** [hotwired/turbo-rails#229]
Any interest in adding a turbo-frame request variant to `Turbo::Frames::FrameRequest`?

```diff
# turbo-rails/app/controllers/turbo/frames/frame_request.rb

module Turbo::Frames::FrameRequest
  extend ActiveSupport::Concern

  included do
    layout -> { false if turbo_frame_request? }
    etag { :frame if turbo_frame_request? }
+   before_action :turbo_frame_request_variant
  end

  private
    def turbo_frame_request?
      request.headers["Turbo-Frame"].present?
    end

+   def turbo_frame_request_variant
+     request.variant = :turbo_frame if turbo_frame_request?
+   end
end
```

This provides the option to change behavior on controller actions for turbo-frame requests:

```ruby
class PostsController < ApplicationController
  def new
    @post = Post.new
    respond_to do |format|
      format.html do |variant|
        variant.turbo_frame # renders app/views/posts/new.html+turbo_frame.erb
        variant.none # renders app/views/posts/new.html.erb
      end
    end
  end

  def create
    respond_to do |format|
      format.turbo_stream do |variant|
        variant.turbo_frame { redirect_to ... } # turbo-frame specific behavior
        variant.none { redirect_to ... }
      end
      format.html { redirect_to ... }
    end
  end
end
```

see also: https://github.com/hotwired/turbo/issues/378

**Comment by WriterZephos** [comment]
I strongly recommend doing this. It doesn't make a lot of sense to not have an easy way to respond differently to from requests coming externally, and requests coming from internal links. For example, if someone types in a url to the show view of some model, you want to render with a template, but if you are responding in the controller to render a partial for internal navigation, they would only get that.

Since I don't like having a bunch of different views that only differ in their file extensions, I like to render the same view and use different templates for replacing turbo frames in the dom or rendering the full page. This works great!

**Comment by WriterZephos** [comment]
Just wanted to drop a link to the gem I am working on that would benefit from this change. The module here dynamically selects a layout based on the request variant:

https://github.com/WriterZephos/TurboRouter/blob/main/lib/turbo_router/controller.rbz

I am rebuilding the above repo using `rails plugin new` instead of `bundle gem` but the demo app that gets generated for testing doesn't have hotwire installed, so I am waiting for that to be fixed.

**Comment by SleeplessByte** [comment]
Even though https://github.com/hotwired/turbo/issues/378#issuecomment-914615544 says it shouldn't, this really should have been solved by registering a new media type, which has the added benefit of working with non-rails solutions, is controllable from the content, and allows you to push the turbo frame as a media type argument.

I do a lot with media types and content negotiations and in my [experienced] opinion tleish isn't correct. The format of the content is indeed HTML, but it's the same issue of `application/json` (syntax) vs `application/vnd.x.y.z+json` which says syntax is `json`, but the type is described more specifically. 

In the same vain, a turbo-frame response actually needs to conform to specific properties (it should contain the same frame) and thus a media type that is an html type is actually appropiate.

The API would then be:

```ruby
format.html do
  # full document response
end
format.turbo_frame do 
  # turbo frame response
end
format.turbo_stream do
  # turbo stream response
end
```

The proposed Accept header for the turbo library would be:

```text
Accept: text/vnd.turbo-frame.html; id=<frame_id>, text/vnd.turbo-stream.html, text/html, application/xhtml+xml
```

Order of accepting is then:
- `text/vnd.turbo-frame.html; id=<frame_id>`
- `text/vnd.turbo-stream.html`
- `text/html`
- `application/xhtml+xml`

So this also means that if you do not want to return a turbo frame:

```ruby
format.html do
  # full document response
end
format.turbo_frame do 
  # turbo frame response
end
```

...and this also means that if the turbo library detects it's not "frameable", it can send:

```text
Accept: text/vnd.turbo-stream.html, text/html, application/xhtml+xml
```

...and you then do not need to do https://github.com/hotwired/turbo-rails/issues/278, because `Accept` should be in the Vary already.

Using a custom header (`Turbo-Frame`) and then using rails specific extra negotiation on the server (variants) makes it harder to interopt with this in other systems.

**Comment by tleish** [comment]
FYI, @dhh does not like the idea of using variants for turbo-frames 

> I don't love the idea of using variants for turbo frames. We did variants originally to provide different types of HTML responses for different platforms, like a native iOS app. Where what you want to send is something substantially different because of the constraints of the platform. I don't think that applies here.
> 
> I'm fine with having a way to detect turbo frame requests as a conditional, but I think that's as far as it should go. This should be the exception. If you're branching all your responses for frames vs not, something isn't right, and we should investigate those pressures in a different way.
> 
> see: https://github.com/hotwired/turbo-rails/pull/250#issuecomment-942079485

**Comment by SleeplessByte** [comment]
Thanks. I had not seen that (yet), but matches how I think about it.

Unfortunately I doubt turbo frames will be turned into a media type.

**Comment by tonywok** [comment]
Curious to hear what you folks think about this use case... I think I agree with DHH that this feels distinct from usage of media type.

I have a GET-form that does a bunch of filtering. I use turbo advance so that folks can share the url and see the filtered results. The response is pretty hefty, so I want to show some kind of loading indicator.

A typical implementation of a turbo frame that lazily sources the frame contents from another controller solves the problem when visiting this page. The contents of the lazy frame can show a loading indicator. However, afaict, this leaves me with the following issues:

1. Upon re-submission of the form (e.g modifying a filter), I don't see the loading indicator (because I'm in the new frame that is not lazy)
2. After advancing the url due to submission, if I were to browser refresh the page, I don't see the spinner because the controller that fields the response returns a non-lazy frame.

I can resolve (1) by using some CSS to hide/show the content and spinner based on the presence of the busy attribute on the turbo frame tag.

I can resolve (2) by using `controller.turbo_frame_request?`, conditionally making the frame lazy if it's _not_ originating from a turbo frame reques -- making it non-lazy if it _is_ originating from a turbo frame request.

**Comment by SleeplessByte** [comment]
I don't think this is "the right" issue for that problem, but it seems that your actually problem (number 2) is that you want a spinner to appear if you're doing a regular browser refresh, which is, in general, a bit weird, but I may not be understanding what you're asking. 

In terms of loading and showing a spinner, you can either:

- Always lazy load it: put the spinner in the logic that starts the lazy load, aka your solution to [1]
- Sometimes lazy load: benefits from the solution above, but regular loading (and no spinner) when doing a _full page reload_.

If you're asking how to do the "always lazy load", there are a lot of solutions to that. In my media-type heavy controllers I don't use `turbo_frame_request?` or anything similar, because many many mediatypes can match on the same controller, but using that to change the response is perfectly fine. Using that to sometimes make it lazy: I guess you could but I don't understand why.

Perhaps GitHub Discussions, twitter, or a new issue is better suited for this discussion @tonywok. If you feel that way, I recommend you don't respond again here (unless with an emoji) but rather open that new issue, and I'll minimize my response to not further pollute this issue.
